messages.go

```diff
@@ -223,6 +230,31 @@ func (n *Node) sendVoteResponse(candidateid string, grant bool) {
 	}
 }
 
+// sendNewQuorumToAll sends a notification to all cluster members and signals a voluntary
+// leave event. Once memberlist processed the leave event internally, this message is used
+// to trigger an immediate change of the new quorum instead of waiting for the dead node to
+// be kicked.
+func (n *Node) sendNewQuorumToAll(newquorum int) {
+	qrmBytes, _ := json.Marshal(NewQuorum{
+		NewQuorum: newquorum,
+	})
+	msgBytes, _ := json.Marshal(Message{
+		Type:    NewQuorumMsg,
+		Content: qrmBytes,
+	})
+
+	for _, member := range n.memberlist.Members() {
+		if member.Name == n.config.ID {
+			continue
+		}
+		if err := n.memberlist.SendReliable(member, msgBytes); err != nil {
+			n.logger.Printf("[ERR] raftify: couldn't send new quorum to %v: %v\n", member.Name, err.Error())
+			continue
```

Comment:
What would happen in this case? Different nodes would have different quorum sizes.

Reply:
Answer by Fabian Bockaj:
If the new quorum can't be sent then the leave message can't be sent either since they are both executed in that order on shutdown. If the leave message can't be sent to the other cluster members then they will suspect the node failed and kick it eventually. So, the cluster essentially falls back to treating the leave as a crash-related event instead of an intentional one and limitation #2 (in the doc) applies, which is...

> There must never fail more than `floor((n-1)/2)` nodes at the same time. Once the failed nodes are kicked out of the memberlist and the size shrinks, the tolerance resets to the new reduced cluster size.

You should definitely check all nodes after a manually triggered leave in order to make sure they have actually received the message. The only thing you have to do if the message gets lost is wait for the node to get kicked by memberlist.

As a rule of thumb: Don't remove more nodes from your cluster than the number of node failures you can tolerate. So, if your cluster has 5 nodes, you shouldn't remove more than 2 nodes at once. You need to wait for the cluster to be resized which also automatically adjusts the quorum to the new size.

The only place this could be a problem is when going from 2 nodes down to 1. In that case, if the message gets lost, the 1 remaining node will not be able to become the leader.